content/post/2019-12-19-mutable-api.Rmarkdown

```diff
@@ -0,0 +1,184 @@
+---
+title: Objects that feel mutable in R
+date: '2019-12-26'
+slug: mutable-api
+tags:
+  - package development
+---
+
+In R, most often, to change an object, you need to re-assign its new value to it. But sometimes, things feel different because objects are mutable or it seems they are, be it in base R code or in the code of packages. Why and how provide a mutable API/interface in R code? In this blog post, we shall explore <s>few odd examples</s> a few examples to better understand mutable APIs in R.
+
+## Preamble: what do we mean by mutable?
+
+### What is mutable
+
+As explained in the [chapter "Names and values" of the Advanced R book by Hadley Wickham](https://adv-r.hadley.nz/names-values.html), in R objects are actually not mutable, because of the [copy-on-modify behaviour](https://adv-r.hadley.nz/names-values.html#copy-on-modify). The book has very clear diagrams showing how a name is bound to a binding corresponding to an object. When you think you're modifying an object, a new object with a new value has been created and bound to the initial name. The original object is unchanged, it's not mutable.
```

Comment:
> in R objects are actually not mutable

Well, some of them are

> because of the [copy-on-modify behaviour

not really _because_ of the behavior, though

> a name is bound to a binding

I think it is better to say that a name is bound to an object

> When you think you're modifying an object, a new object with a new value has been created and bound to the initial name.

well, usually, yes, but some R objects _are_ mutable, and also, optimization can skip the copy. Also, before this, it would be good to have an example for "When you think you're modifying an object", maybe an example of setting a value in a vector or list?